File: frontend/src/api/audit.ts
import { apiClient } from './client';

export type AuditEvent = {
  id: number;
  timestamp: string;
  agent_id?: string | null;
  actor_id?: string | null;
  source_addr?: string | null;
  correlation_id?: string | null;
  operation: string;
  target_type: string;
  target_id?: string | null;
  result: string;
  failure_reason?: string | null;
};

export type AuditEventsResponse = {
  events: AuditEvent[];
};

export async function listGatewayAuditEvents(limit = 100, signal?: AbortSignal): Promise<AuditEventsResponse> {
  return apiClient.request<AuditEventsResponse>(`/api/control-plane/audit?limit=${limit}`, { signal });
}

export async function listAgentAuditEvents(
  agentId: string,
  limit = 100,
  signal?: AbortSignal,
): Promise<AuditEventsResponse> {
  const path = agentId === 'local'
    ? `/api/audit?limit=${limit}`
    : `/api/agents/${encodeURIComponent(agentId)}/audit?limit=${limit}`;
  return apiClient.request<AuditEventsResponse>(path, { signal });
}

File: backend/ic_env_guard/terminal/audit.py
from ic_env_guard.db.audit import AuditEventCreate, AuditRepository


def add_terminal_lifecycle_event(
    repo: AuditRepository,
    operation: str,
    terminal_id: str,
    result: str = "success",
    actor_id: str = "local-admin",
    failure_reason: str | None = None,
) -> None:
    repo.add(
        AuditEventCreate(
            actor_id=actor_id,
            operation=operation,
            target_type="terminal",
            target_id=terminal_id,
            result=result,  # type: ignore[arg-type]
            failure_reason=failure_reason,
        )
    )
